frontend/appflowy_flutter/lib/plugins/database/grid/presentation/grid_page.dart

```diff
@@ -85,67 +93,122 @@ class DesktopGridTabBarBuilderImpl implements DatabaseTabBarItemBuilder {
 }
 
 class GridPage extends StatefulWidget {
-  final DatabaseController databaseController;
   const GridPage({
+    super.key,
     required this.view,
     required this.databaseController,
     this.onDeleted,
-    super.key,
+    this.initialRowId,
   });
 
   final ViewPB view;
+  final DatabaseController databaseController;
   final VoidCallback? onDeleted;
+  final String? initialRowId;
 
   @override
   State<GridPage> createState() => _GridPageState();
 }
 
 class _GridPageState extends State<GridPage> {
-  @override
-  void initState() {
-    super.initState();
-  }
+  bool _didOpenInitialRow = false;
 
   @override
   Widget build(BuildContext context) {
     return MultiBlocProvider(
       providers: [
+        BlocProvider<NotificationActionBloc>.value(
+          value: getIt<NotificationActionBloc>(),
+        ),
         BlocProvider<GridBloc>(
           create: (context) => GridBloc(
             view: widget.view,
             databaseController: widget.databaseController,
           )..add(const GridEvent.initial()),
         ),
       ],
-      child: BlocBuilder<GridBloc, GridState>(
-        builder: (context, state) {
-          return state.loadingState.map(
-            loading: (_) =>
-                const Center(child: CircularProgressIndicator.adaptive()),
-            finish: (result) => result.successOrFail.fold(
-              (_) => GridShortcuts(
-                child: GridPageContent(view: widget.view),
-              ),
-              (err) => FlowyErrorPage.message(
-                err.toString(),
-                howToFix: LocaleKeys.errorDialog_howToFixFallback.tr(),
-              ),
-            ),
-            idle: (_) => const SizedBox.shrink(),
-          );
+      child: BlocListener<NotificationActionBloc, NotificationActionState>(
+        listener: (context, state) {
+          final action = state.action;
+          if (action?.type == ActionType.openRow &&
+              action?.objectId == widget.view.id) {
+            final rowId = action!.arguments?[ActionArgumentKeys.rowId];
+            if (rowId != null) {
+              _openRow(context, rowId);
+            }
+          }
         },
+        child: BlocBuilder<GridBloc, GridState>(
+          builder: (context, state) {
+            return state.loadingState.map(
+              loading: (_) =>
+                  const Center(child: CircularProgressIndicator.adaptive()),
+              finish: (result) => result.successOrFail.fold(
+                (_) {
+                  // Open initial row if any
+                  _openRow(context, widget.initialRowId, true);
+                  return GridShortcuts(
```

Comment:
perform logic here and early return when possible. Reduce responsibility of _openRow to just opening the row. The logic will also be clearer because there's no defaulted parameter

Reply:
Early return? When should there be an early return though? It should always return `GridShortcuts( ... )`

Reply:
Oh, my bad. I still think it would be better to move the logic that determines whether the overlay should be shown out of there though. Does BlocListener work with postframecallback? If not, I would move the checks above the openRow function.